dynamic_api.hpp

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2023 Azul Systems, Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef void signal_handler(int);
+
+extern int checkpoint(const char* const* args, bool stop_current, signal_handler *wrapper, signal_handler **handler_ptr);
+
+typedef void restore_handler();
+
+extern int restore(const char* const* args, restore_handler *on_restore);
+
```

Comment:
I think this should be the same as minicriu-client.h. Restore technically is done by minicriu executable, and not the part of client interface

Reply:
The difference are the handlers. If the minicriu client is loaded statically, and also maybe when there's single thread the API can be simpler (`minicriu_dump()`). Here's the explanation for the presence of signal handler wrappers:
https://github.com/openjdk/crac/compare/crac...rvansa:crac:engine_library#diff-6c90ebfe7b95b2758024613b885a9a709d7c1b5e49d1c9bbbc6bb79e9c4dd552R603

If you're just saying that we should merge `dynamic_api.hpp` into `minicriu-client.h` I don't have an issue with that.

Reply:
I think we need a most generic, single API for static and dynamic cases, but it should be clear. It's not clear how signals are affecting dynamic workflow. Anyway interface may have some extra parameters for static use-case that are not really required.